go/vt/vttablet/tabletmanager/shard_sync.go

```diff
@@ -216,93 +210,75 @@ func (agent *ActionAgent) abortMasterTerm(ctx context.Context, masterAlias *topo
 	log.Infof("Active reparents are enabled; converting MySQL to replica.")
 	demoteMasterCtx, cancelDemoteMaster := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
 	defer cancelDemoteMaster()
-	if _, err := agent.demoteMaster(demoteMasterCtx, false /* revertPartialFailure */); err != nil {
+	if _, err := tm.demoteMaster(demoteMasterCtx, false /* revertPartialFailure */); err != nil {
 		return vterrors.Wrap(err, "failed to demote master")
 	}
 	setMasterCtx, cancelSetMaster := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
 	defer cancelSetMaster()
 	log.Infof("Attempting to reparent self to new master %v.", masterAliasStr)
-	if err := agent.SetMaster(setMasterCtx, masterAlias, 0, "", true); err != nil {
+	if err := tm.SetMaster(setMasterCtx, masterAlias, 0, "", true); err != nil {
 		return vterrors.Wrap(err, "failed to reparent self to new master")
 	}
 	return nil
 }
 
-func (agent *ActionAgent) startShardSync() {
+func (tm *TabletManager) startShardSync() {
 	// Use a buffer size of 1 so we can remember we need to check the state
 	// even if the receiver is busy. We can drop any additional send attempts
 	// if the buffer is full because all we care about is that the receiver will
 	// be told it needs to recheck the state.
-	agent.mutex.Lock()
-	agent._shardSyncChan = make(chan struct{}, 1)
-	agent._shardSyncDone = make(chan struct{})
+	tm.mutex.Lock()
+	defer tm.mutex.Unlock()
+	tm._shardSyncChan = make(chan struct{}, 1)
+	tm._shardSyncDone = make(chan struct{})
 	ctx, cancel := context.WithCancel(context.Background())
-	agent._shardSyncCancel = cancel
-	agent.mutex.Unlock()
-
-	// Queue up a pending notification to force the loop to run once at startup.
-	agent.notifyShardSync()
+	tm._shardSyncCancel = cancel
 
 	// Start the sync loop in the background.
-	go agent.shardSyncLoop(ctx)
+	go tm.shardSyncLoop(ctx, tm._shardSyncChan, tm._shardSyncDone)
+
+	// Queue up a pending notification to force the loop to run once at startup.
+	go tm.notifyShardSync()
 }
 
-func (agent *ActionAgent) stopShardSync() {
+func (tm *TabletManager) stopShardSync() {
 	var doneChan <-chan struct{}
 
-	agent.mutex.Lock()
-	if agent._shardSyncCancel != nil {
-		agent._shardSyncCancel()
-		agent._shardSyncCancel = nil
-		agent._shardSyncChan = nil
-
-		doneChan = agent._shardSyncDone
-		agent._shardSyncDone = nil
+	tm.mutex.Lock()
+	if tm._shardSyncCancel != nil {
+		tm._shardSyncCancel()
```

Comment:
do we no longer need to set `_shardSyncCancel` to nil?

Reply:
Same reason as `close`. The cancel function is idempotent. Multiple calls are harmless.